File: BilisPera/BilisPera/Utils/Location/BPLocationManager.m
//
//  BPLocationManager.m
//  BilisPera
//
//  Created by BHJ on 2025/8/25.
//

#import "BPLocationManager.h"

NS_ASSUME_NONNULL_BEGIN

@interface BPLocationManager ()<CLLocationManagerDelegate>

@property (nonatomic, strong) CLGeocoder *geocoder;
@property (nonatomic, strong) CLLocationManager *manager;

@end

@implementation BPLocationManager

+ (instancetype)shared {
    static BPLocationManager *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[BPLocationManager alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        self.geocoder = [[CLGeocoder alloc] init];
        self.manager = [[CLLocationManager alloc] init];
    }
    return self;
}

- (void)startLocationRequestWithCompletion:(void (^)(void))completion {
    self.manager.delegate = self;
    [self.manager startUpdatingLocation];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        completion();
    });
}

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray<CLLocation *> *)locations {
    CLLocation *location = [locations lastObject];
    [manager stopUpdatingLocation];
    
    BPLocationModel *model = [[BPLocationModel alloc] init];
    model.sighted = [NSString stringWithFormat:@"%f",
                     location.coordinate.latitude];
    model.hills = [NSString stringWithFormat:@"%f",
                   location.coordinate.longitude];
    
    [self saveUserLocationWithModel:model];
    [self reverseGeocodeLocationWithCoordinate:location.coordinate];
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
    [manager stopUpdatingLocation];
}

- (void)reverseGeocodeLocationWithCoordinate:(CLLocationCoordinate2D)coordinate {
    [self.geocoder reverseGeocodeLocation:[[CLLocation alloc] initWithLatitude:coordinate.latitude longitude:coordinate.longitude] completionHandler:^(NSArray<CLPlacemark *> * _Nullable placemarks,
                                                                                                                                                       NSError * _Nullable error) {
        if (placemarks && placemarks.count > 0) {
            CLPlacemark *place = placemarks[0];
            
            BPLocationModel *model = [self getUserLocation];
            if (!model) {
                model = [[BPLocationModel alloc] init];
            }
            model.astonishing = place.ISOcountryCode;
            model.stopped = place.country;
            model.truly = place.administrativeArea ?: (place.locality ?: @"");
            model.theplain = place.locality ?: @"";
            model.horizon = place.name ?: @"";
            [self saveUserLocationWithModel:model];
        }
    }];
}

- (BPLocationModel *)getUserLocation {
    NSDictionary *dic = [[NSUserDefaults standardUserDefaults] dictionaryForKey:@"userLocationKey"];
    BPLocationModel *model = nil;
    if (dic) {
        NSError *error = nil;
        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dic options:0 error:&error];
        if (!error) {
            model = [BPLocationModel mj_objectWithKeyValues:dic];
        } else {
            NSLog(@"转换失败: %@", error);
        }
    }
    return model;
}

- (void)saveUserLocationWithModel:(BPLocationModel *)model {
    NSDictionary *dic = model.mj_keyValues;
    if (dic) {
        [[NSUserDefaults standardUserDefaults] setObject:dic forKey:@"userLocationKey"];
    }
}

@end

NS_ASSUME_NONNULL_END

File: BilisPera/BilisPera/Utils/Location/BPLocationManager.swift
//
//  BPLocationManager.swift
//  BilisPera
//
//  Created by BHJ on 2025/8/20.
//

import Foundation
import CoreLocation
import Combine

@objc
class BPLocationManager: NSObject, @preconcurrency CLLocationManagerDelegate {
    let geocoder = CLGeocoder()
    var manager = CLLocationManager()
    
    @MainActor @objc static let shared = BPLocationManager()
    
    @MainActor
   @objc func startLocationRequest(completion: (() -> Void)? = nil) {
        manager = CLLocationManager()
        manager.delegate = self
        manager.startUpdatingLocation()
        Task{
            try await Task.sleep(nanoseconds: 3 * 1_000_000_000)
            completion?();
        }
    }
    
    @MainActor
    func locationManager(
        _ manager: CLLocationManager,
        didUpdateLocations locations: [CLLocation]
    ) {
        guard let location = locations.last else { return }
        manager.stopUpdatingLocation()
        let model = BPLocationModel.init()
        model.sighted = "\(location.coordinate.latitude)"
        model.hills = "\(location.coordinate.longitude)"
        saveUserLocation(model: model)
        reverseGeocodeLocation(coordinate: location.coordinate)
    }
    
    @MainActor func locationManager(
        _ manager: CLLocationManager,
        didFailWithError error: Error
    ) {
        manager.stopUpdatingLocation()
    }
    
    @MainActor
    func reverseGeocodeLocation(coordinate: CLLocationCoordinate2D){
        geocoder
            .reverseGeocodeLocation(CLLocation.init(latitude: coordinate.latitude, longitude: coordinate.longitude)){[weak self] (
                placemarks,
                error
            ) in
                guard let self,let placemarks = placemarks, let place = placemarks.first else {
                    return
                }
                var model = BPLocationModel()
                if let m = getUserLocation(){
                    model = m;
                }
                model.astonishing = place.isoCountryCode ?? ""
                model.stopped = place.country ?? ""
                model.truly = place.administrativeArea ?? (place.locality ?? "")
                model.theplain = place.locality ?? ""
                model.horizon = place.name ?? ""
                saveUserLocation(model: model)
            }
    }
    
    @objc func getUserLocation() -> BPLocationModel?{
        var model: BPLocationModel?
       if let dic = UserDefaults.standard
        .object(forKey: userLocationKey){
           do {
               let jsonData = try JSONSerialization.data(withJSONObject: dic as Any)
               let decoder = JSONDecoder()
               model = try decoder.decode(BPLocationModel.self, from: jsonData)
           } catch {
               print("转换失败: \(error)")
           }
       }
        return model
    }
    
    private func saveUserLocation(model: BPLocationModel){
        let dic = model.toDictionary()
        UserDefaults.standard.set(dic, forKey: userLocationKey)
    }
}
